src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,296 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Both the documentation and the code is heavily inspired by pyGLMnet:
+ * https://github.com/glm-tools/pyglmnet/
+ *
+ * Author: Tej Sukhatme
+ */
+
+#include <shogun/lib/config.h>
+
+#include <shogun/distributions/Distribution.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/IterativeMachine.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/mathematics/RandomMixin.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/optimization/ConstLearningRate.h>
+#include <shogun/optimization/ElasticNetPenalty.h>
+#include <shogun/optimization/FirstOrderCostFunction.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+
+namespace shogun
+{
+	enum GLM_DISTRIBUTION
+	{
+		POISSON
+	};
+
+	class DotFeatures;
+	class Features;
+	class RegressionLabels;
+	class GLMCostFunction;
+
+	/** @brief Class for estimating regularized generalized linear models (GLM).
+	 *   The regularized GLM minimizes the penalized negative log likelihood:
+	 *
+	 *  This uses Elastic-net penalty which defaults to the ridge penalty when
+	 *  alpha = 0 and defaults to the lasso penalty when alpha = 1.
+	 *
+	 * */
+	class GLM : public RandomMixin<IterativeMachine<LinearMachine>>
+	{
+	public:
+		friend class GLMCostFunction;
+
+		/** default constructor */
+		GLM();
+
+		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda,
+		    float64_t learning_rate, int32_t max_iterations,
+		    float64_t tolerance, float64_t eta);
+
+		/** destructor */
+		~GLM() override
+		{
+		}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels>
+		apply_regression(std::shared_ptr<Features> data = NULL) override;
+
+		/** Returns the name of the SGSerializable instance.  It MUST BE
+		 *  the CLASS NAME without the prefixed `C'.
+		 *
+		 * @return name of the SGSerializable
+		 */
+		const char* get_name() const override
+		{
+			return "GLM";
+		}
+
+		static char* glm_enum_to_string(GLM_DISTRIBUTION distribution)
+		{
+			switch (distribution)
+			{
+			case POISSON:
+				return "Poisson";
+			default:
+				return "Invalid distribution";
+			}
+		}
+
+		virtual EProblemType get_machine_problem_type() const override
+		{
+			return PT_REGRESSION;
+		}
+
+		protected:
+		void init_model(const std::shared_ptr<Features> data);
```

Comment:
```suggestion
		void init_model(const std::shared_ptr<Features>& data);
```

Reply:
I can't do this, this overrides the init_model() from iterative machine and it is done this way there

Reply:
if this method is overriding something you should mark it `override`

Reply:
Yes I'll do that